src/ucp/proto/proto_init.h

```diff
@@ -68,44 +49,40 @@ ucp_proto_perf_envelope_make(const ucs_linear_func_t *funcs, uint64_t funcs_num,
 
 
-/**
- * Initialize the performance of a protocol that consists of several parallel
- * stages. The performance estimations are added to params->caps.
- *
- * @param [in]    proto_name    Protocol name, for debugging.
- * @param [in]    range_start   Range interval start.
- * @param [in]    range_end     Range interval end.
- * @param [in]    bias          Performance bias (0 - no bias).
- * @param [in]    stages        Array of parallel stages performance ranges.
- * @param [in]    num_stages    Number of parallel stages in the protocol.
- * @param [inout] caps          Filled with protocol performance data.
- */
+ucp_proto_common_init_params_t
+ucp_proto_common_params_init(const ucp_proto_init_params_t *init_params);
+
+
 ucs_status_t
-ucp_proto_init_parallel_stages(const char *proto_name, size_t range_start,
-                               size_t range_end, double bias,
-                               const ucp_proto_perf_range_t **stages,
-                               unsigned num_stages, ucp_proto_caps_t *caps);
+ucp_proto_common_add_ppln_perf(ucp_proto_perf_t *perf,
+                               const ucp_proto_perf_segment_t *frag_seg,
+                               size_t max_length);
+
 
+ucs_status_t
+ucp_proto_init_add_memreg_time(const ucp_proto_common_init_params_t *params,
+                               ucp_md_map_t reg_md_map,
+                               ucp_proto_perf_factor_id_t cpu_factor_id,
+                               const char *perf_node_name,
+                               size_t range_start, size_t range_end,
+                               ucp_proto_perf_t *perf);
 
-void ucp_proto_init_memreg_time(const ucp_proto_common_init_params_t *params,
-                                ucp_md_map_t reg_md_map,
-                                ucs_linear_func_t *memreg_time,
-                                ucp_proto_perf_node_t **perf_node_p);
 
 ucs_status_t
-ucp_proto_init_buffer_copy_time(ucp_worker_h worker, const char *title,
-                                ucs_memory_type_t local_mem_type,
-                                ucs_memory_type_t remote_mem_type,
-                                uct_ep_operation_t memtype_op,
-                                ucs_linear_func_t *copy_time,
-                                ucp_proto_perf_node_t **perf_node_p);
+ucp_proto_init_add_buffer_copy_time(ucp_worker_h worker, const char *title,
+                                    ucs_memory_type_t local_mem_type,
+                                    ucs_memory_type_t remote_mem_type,
+                                    uct_ep_operation_t memtype_op,
+                                    size_t range_start, size_t range_end,
+                                    int local, ucp_proto_perf_t *perf);
 
 
 ucs_status_t
-ucp_proto_common_init_caps(const ucp_proto_common_init_params_t *params,
+ucp_proto_common_init_perf(const ucp_proto_common_init_params_t *params,
```

Comment:
It seems this method is also in the wrong place.
It should be `ucp_proto_perf_init`, and moved to `ucp_proto_perf.h`.

Reply:
I am agree that we can consider renaming this function to meet header file prefix (e.g. `ucp_proto_init_perf`) but disagree that it should be part of `proto_perf.h` file. I think that this function on the right place since it contain protocols related logic of extending perf structure .

FMPOV `proto_perf.h` file was designed to contain `ucp_proto_perf_t` structure interfaces which is unrelated to certain protocols logic.

I agree to move `ppln` (your previous comment) logic to `proto_perf.h` file since to implement this logic we need only entities defined in that header file. But implementing `ucp_proto_common_init_perf` requires `ucp_proto_common_init_params_t` and `ucp_proto_common_tl_perf_t` which I am tend consider as protocol-specific stuff.

Reply:
Renamed but left in the `proto_init` file